### PR TITLE
1516 create documentation for invoke scubacached

### LINKED ISSUE #1516: Create documentation for how to use Invoke-ScubaCached
# 💡 Summary #

Invoke-ScubaCached is a secret dev function we've been using since the beginning of SCuBA.

It was used as a way to speed up Rego development by running ScubaGear on a JSON that ScubaGear had already output.
ScubaCached bypassed reaching out to the tenant again and just runs the Rego rules and output the report.
The function was simplistic at first which meant telling people off hand to use it. 

Now it's also being used as part of the functional test orchestrator and being used to test rego on injected settings that can't be tested on a live tenant.  Along with additional parameters to support it's new capabilities 

This issue is to officially document it's capabilities somewhere as it is grown more complex with additional features as time has gone on. 

## Motivation and context ##

Documentation a useful dev function that everyone should be using if they touch Rego. 

## Implementation notes ##

Somewhere in our documentation add a section for 
`How to use Invoke-ScubaCached`

## Acceptance criteria ##

- [ ] **execution.md** file has been updated detailing how `Invoke-ScubaCached` is used and why it's used.

### PR DESCRIPTION
**Pull request by DickTracyII**
Add documentation for Invoke-ScubaCached

## 🗣 Description ##

This PR introduces documentation for the Invoke-ScubaCached command.
The new docs describe what the command does, its parameters, usage scenarios, and how it differs from Invoke-Scuba.

Documentation includes:
- Purpose and functionality of Invoke-ScubaCached.
- Examples of running scans using cached data.
- Guidance on when to use cached vs. fresh scans.
- Notes on limitations and environment considerations.

## 💭 Motivation and context ##

While `Invoke-ScubaCached` has been available as part of ScubaGear for some time, its functionality was not documented.
This created confusion for users on:

- How to reuse cached data effectively.
- Which scenarios benefit from cached scans (e.g., reducing API calls, offline reviews).
- Differences between `Invoke-Scuba` (live scan) and `Invoke-ScubaCached` (cached evaluation).

Adding documentation addresses these gaps and supports user adoption, particularly for environments with strict API throttling or offline validation requirements. This documentation also provided guidance for Developers and those troubleshooting ScubaGear results. 

This resolves [Issue #1516]
Closes #1516

## 🧪 Testing ##

Validation steps performed:

- Reviewed command behavior by running Invoke-ScubaCached against existing cached scan data.
- Verified that documented examples match actual command behavior.
- Confirmed that invalid/missing cache paths trigger the expected error messages.

Steps for reviewers to test the docs:

1. Run a fresh scan with Invoke-Scuba to generate cached data.
2. Execute Invoke-ScubaCached using the cache output.
3. Confirm results match documentation examples.
4. Verify that error handling for invalid cache paths aligns with documented behavior.

## ✅ Pre-approval checklist ##

- [ ] This PR has an informative and human-readable title.
- [ ] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [ ] Changes are limited to a single goal - *eschew scope creep!*
- [ ] Changes are sized such that they do not touch excessive number of files.
- [ ] *All* future TODOs are captured in issues, which are referenced in code comments.
- [ ] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [ ] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [ ] All relevant type-of-change labels added.
- [ ] All relevant project fields are set.
- [ ] All relevant repo and/or project documentation updated to reflect these changes.
- [ ] Unit tests added/updated to cover PowerShell and Rego changes.
- [ ] Functional tests added/updated to cover PowerShell and Rego changes.
- [ ] All relevant functional tests passed.
- [ ] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention
- [ ] Demonstrate changes to the team for questions and comments. 
    (Note: Only required for issues of size `Medium` or larger)

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

File: docs/execution/scubacached.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+- **No credential requirements**: No need for M365 access
+- **Consistent data**: Analysis runs on the same dataset for reproducible results
+
+### Full Mode Benefits
+
+- **Fresh data**: Always uses current tenant configuration
+- **Complete workflow**: Single command for end-to-end assessment
+- **Automatic cleanup**: Handles file management automatically
+
+### Custom Provider File Names
```
Comment (mitchelbaker-cisa) [inline docs/execution/scubacached.md]: The section for "custom provider file names" doesn't make sense under performance considerations. Maybe move it under the usage examples header?

File: docs/execution/scubacached.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+## Version Compatibility
+
+The cached mode works with provider data from the same major version of ScubaGear. When upgrading ScubaGear versions:
+
+- **Minor version updates**: Cached data typically remains compatible
+- **Major version updates**: May require fresh data export due to schema changes
+- **Cross-version analysis**: Not recommended; re-export data with the new version
+
+Check the [release notes](../../RELEASES.md) for version-specific compatibility information.
```
Comment (mitchelbaker-cisa) [inline docs/execution/scubacached.md]:
```suggestion
Check the [release notes](https://github.com/cisagov/ScubaGear/releases) for version-specific compatibility information.
```

**Comment by DickTracyII** [comment]
> Looks good overall. Could we add one more example for how Invoke-SCuBACached could be used when testing Rego on injected settings that can't be tested in a live tenant?
> 
> We could use something like the authentication policy migration settings for AAD 3.4 as an example which @MichaelHicks-MSFT previously worked on in PR [#1778](https://github.com/cisagov/ScubaGear/pull/1778). Testing for this PR involved using Invoke-SCuBACached to modify the JSON output.

@mitchelbaker-cisa I added a section for this. please review. @MichaelHicks-MSFT make sure my example makes sense too

File: docs/execution/scubacached.md (hunk 1 above)
Comment (DickTracyII) [reply docs/execution/scubacached.md]: your right. I removed this. there are plenty of examples that show this already

File: README.md [hunk 3]
```diff
@@ -173,13 +173,16 @@ Invoke-SCuBA -ConfigFilePath "path/to/your/config.yaml" -SilenceBODWarnings
...
 ### Troubleshooting & Support
 
-- [Multiple Tenants](docs/troubleshooting/tenants.md)
-- [Product-Specific Issues](docs/troubleshooting/)
-  - [Defender](docs/troubleshooting/defender.md)
-  - [Exchange Online](docs/troubleshooting/exchange.md)
-  - [Power Platform](docs/troubleshooting/power.md)
-  - [Microsoft Graph](docs/troubleshooting/graph.md)
-- [Network & Proxy](docs/troubleshooting/proxy.md)
+| Topic | Resource | Notes |
```
Comment (mitchelbaker-cisa) [inline README.md]: I like the concept of a table but for the nested product-specific issues the format is confusing. Can we switch back to the list format? 

Image

```
- [Multiple Tenants](docs/troubleshooting/tenants.md)
- [Product-Specific Issues](docs/troubleshooting/)
  - [Defender](docs/troubleshooting/defender.md)
  - [Exchange Online](docs/troubleshooting/exchange.md)
  - [Power Platform](docs/troubleshooting/power.md)
  - [Microsoft Graph](docs/troubleshooting/graph.md)
- [Network & Proxy](docs/troubleshooting/proxy.md)
- [Cached Execution](docs/execution/scubacached.md)
```

File: docs/execution/scubacached.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+For example, when testing AAD authentication policy migration settings
+
+1. Export provider data in full mode:
+
+```powershell
+Invoke-SCuBACached -ProductNames aad -OutPath "C:\ScubaResults\AAD_Policy_Test"
+```
+
+2. Manually modify the `ProviderSettingsExport.json` file to inject test scenarios (example):
```
Comment (mitchelbaker-cisa) [inline docs/execution/scubacached.md]: While `ProviderSettingsExport.json` can be used, it is the legacy file option. Switch this to `ScubaResults_<guid>.json` instead.

File: docs/execution/scubacached.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+```powershell
+Invoke-SCuBACached -ProductNames exo, defender -ExportProvider $false -OutPath "C:\ScubaResults\M365BaselineConformance_2025_09_22_10_19_24"
+```
+
+### Testing Rego on Modified Settings
+
+When testing Rego policies against specific configuration scenarios that can't be replicated in a live tenant, you can modify the exported JSON data and use cached mode to test the policy logic.
+
+For example, when testing AAD authentication policy migration settings
```
Comment (mitchelbaker-cisa) [inline docs/execution/scubacached.md:163]: We could define a couple different scenarios here. 

```
For example, when testing AAD authentication policy migration settings:

// Update the "PolicyMigrationState" property to test with other values such as "preMigration", "migrationComplete", or "migrationInProgress"
{
  "authentication_method_policy":  {
    "AuthenticationMethodConfigurations@odata.context":  "https://graph.microsoft.com/beta/$metadata#policies/authenticationMethodsPolicy/authenticationMethodConfigurations",
    "PolicyVersion":  "1.5",
    "PolicyMigrationState":  null,
    "DisplayName":  "Authentication Methods Policy"
  }
}
```

```
Or when testing AAD conditional access policy settings:

// Update the "State" property to test with other values such as "Report only" or "disabled"
{
  "conditional_access_policies": [
    {
      "DisplayName": "Test Migration Policy",
      "State": "enabled",
      "Conditions": {
        "Users": {
          "IncludeUsers": ["All"]
        }
      },
      "GrantControls": {
        "BuiltInControls": ["mfa"]
      }
    }
  ]
}
```

File: docs/execution/scubacached.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
+# ScubaGear Cached Execution
+
+`Invoke-SCuBACached` is a specialized version of the main ScubaGear function that allows you to run security assessments in two distinct modes:
+
+1. **Full mode** - Works identically to `Invoke-SCuBA` (exports provider data, runs Rego analysis, and generates reports)
```
Comment (buidav) [inline docs/execution/scubacached.md]: Not exactly. One thing `Invoke-Scubacached` does different than `Invoke-SCuBA` even when run regularly is how the output artifacts are grouped together. 

When `Invoke-SCuBA` is executed a directory folder of the form `M365BaselinConformance_{datetimestamp}` is generated in the file path specified by the `OutPath` parameter.  The ScubaGear outputs (json, html) files are put under this folder.

When `Invoke-ScubaCached` is executed the directory folder (M365BaselineConformance) is NOT created and the SCuBA outputs (JSON, HTML files etc.)  are put directly into the file path specified by `OutPath`.

The fact a M365BaselineConformance folder is not created by Invoke-ScubaCached was it to make it easier to do rego develpment without having to rerun the providers.
Run full mode once to pull in a provider JSON.
Then on each subsequent run just run in cached mode. The user doesn't need to mess with `OutPath`.

File: docs/execution/scubacached.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+- **Complete workflow**: Single command for end-to-end assessment
+- **Automatic cleanup**: Handles file management automatically
+
+## Best Practices
+
+### Development and Testing
+
+- Use cached mode during development to speed up iteration cycles
+- Keep a "golden" export for consistent testing across different configurations
+- Use descriptive output folder names to organize different test scenarios
```
Comment (buidav) [inline docs/execution/scubacached.md:323]: Specifically for Rego development. This was what the function was orginally made for. 
A developer would be able to work on the rego then run `ScubaCached` to quickly see their rego changes on the HTML report. Without having to wait for the entire provider to run.


Our own main use case for this this function right now is the [functional test pipeline](https://github.com/cisagov/ScubaGear/actions/workflows/test_production_function.yaml). The code for the pipeline makes HEAVY use of this function to manipulate cached JSON to test SCuBA baseline policies that can't be tested safely. Such as manipulating MFA conditional access policies.

**Review by buidav — CHANGES_REQUESTED** [review]

File: docs/execution/scubacached.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+
+This approach allows you to test edge cases, policy combinations, or specific configurations that would be difficult or impossible to create in a live Microsoft 365 environment.
+
+#### EXAMPLE 2
+
+Testing if the tenant had the migration complete configured. Newer tenants do not have this option. To test this, you can change the results json to simulate an alternate pass. For exmample change:
+
+```json
+"authentication_method_policy":  {
+    "AuthenticationMethodConfigurations@odata.context":  "https://graph.microsoft.com/beta/$metadata#policies/authenticationMethodsPolicy/authenticationMethodConfigurations%22,
```
Comment (mitchelbaker-cisa) [inline docs/execution/scubacached.md]: The values for "AuthenticationMethodConfiguration" and "PolicyMigrationState" need to be strings.

Image

File: README.md (hunk 3 above)
Comment (DickTracyII) [reply README.md]: Yeah we can. I felt the table would be nice if we add a note to what each doc was for, providing the user with a "reason" why these would be used. Otherwise, the users will be all over the place, looking for an answer.

File: docs/execution/scubacached.md (hunk 4 above)
Comment (DickTracyII) [reply docs/execution/scubacached.md]: updated

File: docs/execution/scubacached.md (hunk 6 above)
Comment (DickTracyII) [reply docs/execution/scubacached.md]: hopefully I captured what you were saying in the new update.

File: docs/execution/scubacached.md (hunk 7 above)
Comment (DickTracyII) [reply docs/execution/scubacached.md:323]: @buidav i changed this a little capture what you said. does this work?

File: docs/execution/scubacached.md (hunk 8 above)
Comment (DickTracyII) [reply docs/execution/scubacached.md]: fixed. good catch
